<h2 align="center" style="color:#0d47a1; font-family:Georgia;">
    GUÍA DE LABORATORIO: <br>Álgebra lineal para Machine Learning, escrita en NumPy
</h2>
<h3 align="center" style="font-family:Trebuchet MS;">
    Nombre: Roberto Roly Copa Mamani
</h3>
<h3 align="center" style="font-family:Trebuchet MS;">
    Materia: Machine Learning &nbsp;&nbsp; INF 672
</h3>

In [2]:
import numpy as np
from sklearn.datasets import load_iris

# Representación de datos y selección de muestras
iris = load_iris()
X = iris.data
y = iris.target

# Selección de muestras:
# Índice 0: Setosa (clase 0)
# Índice 1: Setosa (clase 0) - Misma especie que idx 0
# Índice 50: Versicolor (clase 1) - Distinta especie
idx_set1, idx_set2, idx_ver = 0, 1, 50

muestras = np.array([X[idx_set1], X[idx_set2], X[idx_ver]])
etiquetas_nombres = ['Setosa_1 (idx 0)', 'Setosa_2 (idx 1)', 'Versicolor (idx 50)']

# Verificación de producto punto a mano vs NumPy
a = X[idx_set1]
b = X[idx_ver]
dot_manual = sum(ai * bi for ai, bi in zip(a, b))
dot_numpy = np.dot(a, b)

print("*** Producto Punto (a mano vs NumPy) ***")
print(f"A mano: {dot_manual:.2f} | np.dot: {dot_numpy:.2f}\n")

*** Producto Punto (a mano vs NumPy) ***
A mano: 53.76 | np.dot: 53.76



In [4]:
# Matriz 3x3 de Distancias Euclidianas
matriz_distancias = np.zeros((3, 3))
for i in range(3):
    for j in range(3):
        # Distancia euclidiana
        matriz_distancias[i, j] = np.linalg.norm(muestras[i] - muestras[j])

print("*** Matriz 3x3 de Distancias Euclidianas ***")
print(np.round(matriz_distancias, 4))
print()

*** Matriz 3x3 de Distancias Euclidianas ***
[[0.     0.5385 4.0037]
 [0.5385 0.     4.0963]
 [4.0037 4.0963 0.    ]]



In [5]:
# Matriz 3x3 de Similitud Coseno
def similitud_coseno(u, v):
    return np.dot(u, v) / (np.linalg.norm(u) * np.linalg.norm(v))

matriz_coseno = np.zeros((3, 3))
for i in range(3):
    for j in range(3):
        matriz_coseno[i, j] = similitud_coseno(muestras[i], muestras[j])

print("*** Matriz 3x3 de Similitud Coseno ***")
print(np.round(matriz_coseno, 4))
print()

*** Matriz 3x3 de Similitud Coseno ***
[[1.     0.9986 0.9284]
 [0.9986 1.     0.94  ]
 [0.9284 0.94   1.    ]]



In [7]:
# Verificación de ||x||² = x · x
x = X[idx_set1]
norma_cuadrada = np.linalg.norm(x) ** 2
producto_punto_propio = np.dot(x, x)

print("*** Verificación ||x||² = x · x ***")
print(f"||x||² : {norma_cuadrada:.4f}")
print(f"x · x  : {producto_punto_propio:.4f}")

*** Verificación ||x||² = x · x ***
||x||² : 40.2600
x · x  : 40.2600


### 5. Interpretación de resultados

* **Par más parecido según coseno:** El par Setosa_1 y Setosa_2 (índices 0 y 1) es el más parecido, alcanzando una similitud coseno de 0.9984 (muy cercana a 1).
* **Coincidencia con distancia:** Sí coincide, ya que el par Setosa_1 y Setosa_2 registra la distancia euclidiana más pequeña (0.5385), en comparación con las distancias hacia Versicolor (> 4.0).
* **¿Por qué?:** Ambas métricas reflejan que las muestras de la misma especie (Setosa) comparten proporciones e intensidades de características casi idénticas en el espacio vectorial. La distancia euclidiana mide la cercanía física entre los puntos, mientras que la similitud coseno evalúa la orientación de sus vectores; al pertenecer a la misma clase, apuntan prácticamente en la misma dirección y se ubican muy cerca en el espacio.